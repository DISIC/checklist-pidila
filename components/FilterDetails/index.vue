<template>
  <p class="filters-recap alert alert-info" v-html="displayFiltersRecap" aria-live="assertive" aria-atomic="true" />
</template>
<script>
  import FilterConfig from '../../data/filter.config'

  const DEFAULT_OPERATOR = ''
  const TAG_CONNECTOR = ' ou '
  const NO_FILTER_MESSAGE = 'sans aucun filtre ni recherche appliqués pour le moment.'

  export default {
    name: 'FilterDetails',
    computed: {
      numberOfCriteriasDisplayed () {
        return this.$store.getters.displayedCriteriasNumber
      },
      displayActiveFilters () {
        const orderedFilters = this.$store.getters.orderedFilters
        const numberOfCheckedTags = orderedFilters.reduce((sumOfTags, filter) => sumOfTags + filter.tags.length, 0)

        if (numberOfCheckedTags === 0) {
          return ''
        }

        return orderedFilters.reduce((displayedMessage, currentFilter) => {
          if (currentFilter.tags.length === 0) {
            return displayedMessage
          }

          const operator = FilterConfig.operators[currentFilter.title] || DEFAULT_OPERATOR
          const tagString = currentFilter.tags
            .map(tag => '<b>' + tag + '</b>')
            .join(TAG_CONNECTOR)

          return [displayedMessage, operator, tagString].join(' ')
        }, '').trim()
      },
      displayCriteriaText () {
        const criteriaTrad = {
          singular: 'critère',
          plural: 'critères'
        }
        let displayedMessage = ''
        if (this.numberOfCriteriasDisplayed <= 1) {
          displayedMessage = criteriaTrad.singular
        } else {
          displayedMessage = criteriaTrad.plural
        }
        return displayedMessage
      },
      displaySearchQuery () {
        const searchQuery = this.$store.state.searchQuery
        return searchQuery.length > 0 ? `correspondant à la recherche « <b>${searchQuery}</b> »` : ''
      },
      displayFiltersRecap () {
        let filterMessage = NO_FILTER_MESSAGE
        if (this.displaySearchQuery !== '' || this.displayActiveFilters !== '') {
          filterMessage = this.displaySearchQuery + ' ' + this.displayActiveFilters + ' <a href="./" class="filters-reset">Annuler tous les filtres</a>'
        }
        return '<span class="row-count">' + this.numberOfCriteriasDisplayed + '</span> ' +
          this.displayCriteriaText + ' ' + filterMessage
      }
    }
  }

</script>
