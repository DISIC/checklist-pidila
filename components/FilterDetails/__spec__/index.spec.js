import {shallowMount} from '@vue/test-utils'
import FilterConfig from '../../../data/filter.config'
import FilterDetails from '../index.vue'

describe('components | FilterDetails', () => {
  let $store = {}
  beforeEach(() => {
    $store = {
      getters: {
        displayedCriteriasNumber: 1,
        orderedFilters: []
      },
      state: {
        searchQuery: ''
      }
    }
    FilterConfig.operators = {
      'Profil': 'pour',
      'Référentiel': 'dans',
      'Niveau': 'de niveau',
      'Thématique': 'concernant'
    }
  })

  it('should match snapshot', () => {
    // when
    const wrapper = shallowMount(FilterDetails, {mocks: {$store}})
    // then
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should update the row-count span when the component is mounted', () => {
    // when
    const wrapper = shallowMount(FilterDetails, {mocks: {$store}})
    // then
    expect(wrapper.find('span.row-count').element.textContent).toBe('1')
  })

  it('should update the row-count span when the getter from store is updated', () => {
    // when
    const wrapper = shallowMount(FilterDetails, {mocks: {$store}})
    $store.getters.displayedCriteriasNumber = 3

    // then
    expect(wrapper.find('span.row-count').element.textContent).toBe('3')
  })

  it('should update the filters-recap p with "aucun filtre ni recherche appliqués pour le moment" ' +
    'when store has no checked filters and no search query', () => {
    // given
    const expectedNoFilterHtml = '<span class="row-count">1</span>' +
      ' critère sans aucun filtre ni recherche appliqués pour le moment.'

    // when
    const wrapper = shallowMount(FilterDetails, {mocks: {$store}})

    // then
    expect(wrapper.find('.filters-recap').element.innerHTML).toBe(expectedNoFilterHtml)
  })

  it('should update the filters-recap p with "Aucun filtre n\'a été appliqué pour le moment." ' +
    'when no tags are checked but there are empty filters in store and no search query', () => {
    // given
    $store.getters.orderedFilters = [
      {
        title: 'Profil',
        tags: []
      },
      {
        title: 'Référentiel',
        tags: []
      }
    ]
    const expectedNoFilterHtml = '<span class="row-count">1</span>' +
      ' critère sans aucun filtre ni recherche appliqués pour le moment.'

    // when
    const wrapper = shallowMount(FilterDetails, {mocks: {$store}})

    // then
    expect(wrapper.find('.filters-recap').element.innerHTML).toBe(expectedNoFilterHtml)
  })

  it('should update the filters-recap p if there are filters and a search query', () => {
    // Given
    $store.getters.orderedFilters = [
      {
        title: 'Profil',
        tags: ['Intégration']
      },
      {
        title: 'Référentiel',
        tags: ['RGAA', 'Bonnes pratiques Opquast']
      },
      {
        title: 'Niveau',
        tags: []
      },
      {
        title: 'Thématique',
        tags: ['Présentation']
      }
    ]
    $store.state.searchQuery = 'Navigation'
    const expectedNoFilterHtml = '<span class="row-count">1</span>' +
      ' critère correspondant à la recherche « <b>Navigation</b> » pour <b>Intégration</b> dans <b>RGAA</b> ou ' +
      '<b>Bonnes pratiques Opquast</b> concernant <b>Présentation</b> ' +
      '<a href="./" class="filters-reset">Annuler tous les filtres</a>'

    // When
    const wrapper = shallowMount(FilterDetails, {mocks: {$store}})

    // Then
    const trimmedFilterRecapHtml = wrapper.find('.filters-recap').element.innerHTML
      .split('\n').map(HTMLline => HTMLline.trim()).join('\n')
    expect(trimmedFilterRecapHtml).toBe(expectedNoFilterHtml)
  })

  it('should update the filters-recap p if there is a search a query and no filters', () => {
    // Given
    $store.getters.orderedFilters = []
    $store.state.searchQuery = 'Navigation'
    const expectedNoFilterHtml = '<span class="row-count">1</span>' +
      ' critère correspondant à la recherche « <b>Navigation</b> »  ' +
      '<a href="./" class="filters-reset">Annuler tous les filtres</a>'

    // When
    const wrapper = shallowMount(FilterDetails, {mocks: {$store}})

    // Then
    const trimmedFilterRecapHtml = wrapper.find('.filters-recap').element.innerHTML
      .split('\n').map(HTMLline => HTMLline.trim()).join('\n')
    expect(trimmedFilterRecapHtml).toBe(expectedNoFilterHtml)
  })

  it('should display "critère" if 0 or 1 criterias are displayed', () => {
    // given
    $store.getters.displayedCriteriasNumber = 1

    // when
    const wrapper = shallowMount(FilterDetails, {mocks: {$store}})

    // then
    expect(wrapper.element.innerHTML.match(/ critère(?!s)/)).toHaveLength(1)
  })
})
