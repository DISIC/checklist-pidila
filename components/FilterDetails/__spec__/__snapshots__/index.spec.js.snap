// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components | FilterDetails should match snapshot 1`] = `
<p
  aria-atomic="true"
  aria-live="assertive"
  class="filters-recap alert alert-info"
>
  <span
    class="row-count"
  >
    1
  </span>
   critère sans aucun filtre ni recherche appliqués pour le moment.
</p>
`;
